File: packages/core-py/src/mcp_core/auth/relay_login.py
"""MCP relay edge auth — password gate for ``/authorize`` (Python parity).

When ``MCP_RELAY_PASSWORD`` is set, the ``/authorize`` route is fronted by a
thin cookie-session check. Unauthenticated requests are redirected to
``/login`` where the user submits the password once, gets a 24h
``mcp_relay_session`` cookie, and continues to ``/authorize``.

Empty / unset password disables the gate entirely (single-user dev mode).

Brute-force protection: 5 wrong submissions per IP within a 15-minute
sliding window block further attempts with HTTP 429 + ``Retry-After``.

Per spec ``2026-05-01-stdio-pure-http-multiuser §4.2.1``. TypeScript parity
lives at ``packages/core-ts/src/auth/relay-login.ts``.
"""

from __future__ import annotations

import html
import hmac
import secrets
import time
from typing import Any
from urllib.parse import quote

from starlette.responses import HTMLResponse, RedirectResponse, Response

from mcp_core.auth.credential_form import render_form_shell

# In-memory session + brute-force stores. Module-scope is intentional: each
# container hosts a single relay app, so a single shared map matches the
# operational model. Test code resets these directly via fixtures.
_sessions: dict[str, float] = {}  # sid -> expires_at_epoch
_fails: dict[str, tuple[int, float]] = {}  # ip -> (count, first_at_epoch)

SESSION_TTL_S = 24 * 60 * 60
FAIL_WINDOW_S = 15 * 60
FAIL_LIMIT = 5

_configured_password = ""


def configure_relay_login(password: str) -> None:
    """Set the password the gate should enforce. Empty string disables."""
    global _configured_password
    _configured_password = password or ""


def _timing_safe_equal(a: bytes, b: bytes) -> bool:
    """Compare two byte strings safely, mitigating length-leaking timing attacks.

    hmac.compare_digest returns early if lengths differ, which leaks the
    length of the secret. This ensures compare_digest is always called with
    equal-length inputs, hiding the true length of the secret.
    """
    is_length_equal = len(a) == len(b)
    compare_b = b if is_length_equal else a
    return hmac.compare_digest(a, compare_b) and is_length_equal


def _bump_fail(ip: str) -> tuple[bool, int]:
    """Record a failed login attempt; return ``(blocked, retry_after_s)``."""
    now = time.time()
    entry = _fails.get(ip)
    if not entry or now - entry[1] > FAIL_WINDOW_S:
        _fails[ip] = (1, now)
        return False, 0
    count, first_at = entry
    count += 1
    _fails[ip] = (count, first_at)
    if count > FAIL_LIMIT:
        return True, int(FAIL_WINDOW_S - (now - first_at))
    return False, 0


def _clear_fail(ip: str) -> None:
    _fails.pop(ip, None)


def _get_safe_next(input_val: Any) -> str:
    """Validates the ``next`` parameter to prevent Open Redirect vulnerabilities.

    It must start with a single ``/`` and NOT be followed by another ``/``,
    ``\\``, or any whitespace/control character that some browsers might
    normalize into a protocol-relative URL.
    """
    next_ = str(input_val or "/authorize")
    if not next_.startswith("/") or next_.startswith("//") or next_.startswith("/\\") or next_.startswith("\\\\"):
        return "/authorize"
    # Block cases like "/ google.com" or "/\tgoogle.com" (ASCII <= 32).
    if len(next_) > 1 and ord(next_[1]) <= 32:
        return "/authorize"
    return next_


async def require_relay_session(
    cookies: dict[str, str],
    original_url: str,
    password: str | None = None,
) -> RedirectResponse | None:
    """Cookie gate check.

    Returns ``None`` when the request may proceed (gate disabled or cookie
    valid). Otherwise returns a 302 redirect to ``/login?next=<encoded>``.
    """
    pwd = password if password is not None else _configured_password
    if not pwd:
        return None
    sid = cookies.get("mcp_relay_session")
    if sid and _sessions.get(sid, 0) > time.time():
        return None
    next_ = quote(original_url, safe="")
    return RedirectResponse(url=f"/login?next={next_}", status_code=302)


def _render_login_form(next_url: str, error_msg: str | None = None) -> str:
    safe_next = html.escape(str(next_url))
    error_html = ""
    aria_attributes = ' aria-describedby="relay-desc"'
    if error_msg:
        error_html = f"""
            <div id="login-error" class="status-box error" role="alert" style="display: block; margin-bottom: 1.25rem; margin-top: 0;">
                {html.escape(error_msg)}
            </div>"""
        aria_attributes = (
            ' aria-invalid="true" aria-errormessage="login-error" aria-describedby="login-error relay-desc"'
        )

    return f'''    <div class="container">
        <div class="card">
            <div class="server-header">
                <h1 class="server-name">Relay login</h1>
                <div class="server-id">mcp-relay</div>
                <p class="server-description" id="relay-desc">Enter the relay password shared by your deployer.</p>
            </div>{error_html}

            <p class="form-title" id="form-title">Authenticate</p>

            <form method="POST" action="/login" aria-labelledby="form-title" novalidate>
                <!-- nosemgrep: python.fastapi.security.injection.raw-html-format.raw-html-format -- safe_next is escaped -->
                <input type="hidden" name="next" value="{safe_next}">
                <div class="field-group">
                    <label for="field-password" class="field-label">
                        Relay password
                        <span class="required-badge" aria-hidden="true">Required</span>
                    </label>
                    <input
                        id="field-password"
                        type="password"
                        name="password"
                        class="field-input"
                        placeholder="Relay password"
                        autocomplete="current-password"
                        autocorrect="off"
                        autocapitalize="off"
                        spellcheck="false"
                        required
                        autofocus{aria_attributes}
                    />
                </div>

                <button type="submit" class="submit-btn">Continue</button>
            </form>
        </div>
    </div>'''


async def login_get_handler(next: str = "/authorize") -> HTMLResponse:
    """Render the password form using the shared relay form shell.

    The visible card mirrors the relay credential form: dark theme, Inter
    fall-back font stack, ``.field-group`` / ``.field-label`` / ``.field-input``
    classes for the password input, and a primary submit button matching the
    "Connect" styling. The form keeps a plain HTML POST (no JavaScript) so the
    gate works even with a strict CSP that blocks inline scripts.
    """
    next = _get_safe_next(next)
    return HTMLResponse(render_form_shell("Relay login", _render_login_form(next)), headers={"X-Frame-Options": "DENY"})


async def login_post_handler(form: dict, ip: str) -> Response:
    """Verify the password and (on success) issue a session cookie."""
    now = time.time()
    fail = _fails.get(ip)
    if fail and now - fail[1] < FAIL_WINDOW_S and fail[0] >= FAIL_LIMIT:
        retry_after = int(FAIL_WINDOW_S - (now - fail[1]))
        return Response(
            "Too many login attempts. Try again later.",
            status_code=429,
            headers={"Retry-After": str(retry_after)},
        )
    password = str(form.get("password", ""))
    next_ = _get_safe_next(form.get("next"))
    if not _configured_password or not _timing_safe_equal(password.encode(), _configured_password.encode()):
        _bump_fail(ip)
        return HTMLResponse(
            render_form_shell("Relay login", _render_login_form(next_, "Invalid password. Please try again.")),
            status_code=401,
            headers={"X-Frame-Options": "DENY"},
        )
    _clear_fail(ip)
    sid = secrets.token_hex(32)
    _sessions[sid] = time.time() + SESSION_TTL_S
    cookie = f"mcp_relay_session={sid}; HttpOnly; Secure; SameSite=Lax; Max-Age={SESSION_TTL_S}; Path=/"
    return RedirectResponse(url=next_, status_code=302, headers={"set-cookie": cookie})

File: packages/core-ts/src/auth/relay-login.ts
/**
 * MCP relay edge auth — password gate for ``/authorize``.
 *
 * When ``MCP_RELAY_PASSWORD`` is set, the ``/authorize`` route is fronted by
 * a thin cookie-session middleware. Unauthenticated requests are redirected
 * to ``/login`` where the user submits the password once, gets a 24h
 * ``mcp_relay_session`` cookie, and continues to ``/authorize``.
 *
 * Empty / unset password disables the gate entirely (single-user dev mode).
 *
 * Brute-force protection: 5 wrong submissions per IP within a 15-minute
 * sliding window blocks further attempts with HTTP 429 + ``Retry-After``.
 *
 * Per spec ``2026-05-01-stdio-pure-http-multiuser §4.2.1``. Python parity
 * lives at ``packages/core-py/src/mcp_core/auth/relay_login.py``.
 */
import crypto from 'node:crypto'
import { renderFormShell } from './credential-form.js'

interface SessionEntry {
  expiresAt: number
}
interface FailEntry {
  count: number
  firstAt: number
}

const sessions = new Map<string, SessionEntry>()
const fails = new Map<string, FailEntry>()
const SESSION_TTL_MS = 24 * 60 * 60 * 1000
const FAIL_WINDOW_MS = 15 * 60 * 1000
const FAIL_LIMIT = 5

let configuredPassword = ''

export function configureRelayLogin(password: string): void {
  configuredPassword = password ?? ''
}

/**
 * Reset module-scoped state. TEST-ONLY — reset between cases so brute-force
 * counters from one test don't bleed into the next.
 */
export function __resetRelayLoginState(): void {
  sessions.clear()
  fails.clear()
  configuredPassword = ''
}

function timingSafeEqual(a: string, b: string): boolean {
  const ab = Buffer.from(a)
  const bb = Buffer.from(b)
  const isLengthEqual = ab.length === bb.length
  const compareB = isLengthEqual ? bb : ab
  return crypto.timingSafeEqual(ab, compareB) && isLengthEqual
}

function bumpFail(ip: string): { blocked: boolean; retryAfter: number } {
  const now = Date.now()
  const entry = fails.get(ip)
  if (!entry || now - entry.firstAt > FAIL_WINDOW_MS) {
    fails.set(ip, { count: 1, firstAt: now })
    return { blocked: false, retryAfter: 0 }
  }
  entry.count += 1
  if (entry.count > FAIL_LIMIT) {
    return {
      blocked: true,
      retryAfter: Math.ceil((FAIL_WINDOW_MS - (now - entry.firstAt)) / 1000)
    }
  }
  return { blocked: false, retryAfter: 0 }
}

function clearFail(ip: string): void {
  fails.delete(ip)
}

/**
 * Minimal req/res shapes the handlers + middleware need. We accept the
 * Express-style fields directly so consumers can adapt the project's
 * IncomingMessage / ServerResponse pair (or the Express runtime, or test
 * stubs) without dragging the full DOM types into this module.
 */
export interface RelayLoginRequest {
  cookies?: Record<string, string | undefined>
  originalUrl?: string
  query?: Record<string, string | undefined>
  body?: Record<string, string | undefined>
  ip?: string
}

export interface RelayCookieOptions {
  httpOnly?: boolean
  secure?: boolean
  sameSite?: 'lax' | 'strict' | 'none'
  maxAge?: number
}

export interface RelayLoginResponse {
  redirect: (url: string) => unknown
  status: (code: number) => RelayLoginResponse
  send: (body?: unknown) => unknown
  cookie: (name: string, value: string, options?: RelayCookieOptions) => unknown
  header?: (name: string, value: string) => unknown
  set?: (name: string, value: string) => unknown
}

/**
 * Express-style middleware adapter.
 *
 * Pass-through when no password configured. Otherwise: read
 * ``mcp_relay_session`` cookie, allow if present + unexpired, redirect to
 * ``/login?next=<encoded>`` otherwise.
 */
export function createRelayLoginMiddleware(opts: { password: string }) {
  return async (req: RelayLoginRequest, res: Pick<RelayLoginResponse, 'redirect'>, next: () => void): Promise<void> => {
    if (!opts.password) {
      next()
      return
    }
    const sid = req.cookies?.mcp_relay_session
    const entry = sid ? sessions.get(sid) : undefined
    if (entry && entry.expiresAt > Date.now()) {
      next()
      return
    }
    const next_ = encodeURIComponent(req.originalUrl ?? '/authorize')
    res.redirect(`/login?next=${next_}`)
  }
}

function escapeHtml(s: string): string {
  return s.replace(/[&<>"']/g, (c) => {
    const entities: Record<string, string> = {
      '&': '&amp;',
      '<': '&lt;',
      '>': '&gt;',
      '"': '&quot;',
      "'": '&#39;'
    }
    return entities[c] ?? c
  })
}

/**
 * Validates the ``next`` parameter to prevent Open Redirect vulnerabilities.
 *
 * It must start with a single ``/`` and NOT be followed by another ``/``,
 * ``\\``, or any whitespace/control character that some browsers might
 * normalize into a protocol-relative URL.
 */
function getSafeNext(input: unknown): string {
  const next = String(input ?? '/authorize')
  if (!next.startsWith('/') || next.startsWith('//') || next.startsWith('/\\') || next.startsWith('\\\\')) {
    return '/authorize'
  }
  // Block cases like "/ google.com" or "/\tgoogle.com" (ASCII <= 32).
  if (next.length > 1 && next.charCodeAt(1) <= 32) {
    return '/authorize'
  }
  return next
}

function renderLoginForm(next: string, errorMsg?: string): string {
  const safeNext = escapeHtml(String(next))
  const errorHtml = errorMsg
    ? `\n            <div id="login-error" class="status-box error" role="alert" style="display: block; margin-bottom: 1.25rem; margin-top: 0;">\n                ${escapeHtml(errorMsg)}\n            </div>`
    : ''

  const ariaAttributes = errorMsg ? ' aria-invalid="true" aria-errormessage="login-error"' : ''

  return `    <div class="container">
        <div class="card">
            <div class="server-header">
                <h1 class="server-name">Relay login</h1>
                <div class="server-id">mcp-relay</div>
                <p class="server-description">Enter the relay password shared by your deployer.</p>
            </div>${errorHtml}

            <p class="form-title" id="form-title">Authenticate</p>

            <form method="POST" action="/login" aria-labelledby="form-title">
                <!-- nosemgrep: javascript.express.security.injection.raw-html-format.raw-html-format -- safeNext is escapeHtml(next) -->
                <input type="hidden" name="next" value="${safeNext}">
                <div class="field-group">
                    <label for="field-password" class="field-label">
                        Relay password
                        <span class="required-badge" aria-hidden="true">Required</span>
                    </label>
                    <input
                        id="field-password"
                        type="password"
                        name="password"
                        class="field-input"
                        placeholder="Relay password"
                        autocomplete="current-password"
                        autocorrect="off"
                        autocapitalize="off"
                        spellcheck="false"
                        required
                        autofocus${ariaAttributes}
                    />
                </div>

                <button type="submit" class="submit-btn">Continue</button>
            </form>
        </div>
    </div>`
}

export async function loginGetHandler(
  req: Pick<RelayLoginRequest, 'query'>,
  res: Pick<RelayLoginResponse, 'send' | 'set'>
): Promise<void> {
  const next = getSafeNext(req.query?.next)
  res.set?.('Content-Type', 'text/html')
  res.send(renderFormShell('Relay login', renderLoginForm(next)))
}

export async function loginPostHandler(
  req: Pick<RelayLoginRequest, 'body' | 'ip'>,
  res: RelayLoginResponse
): Promise<void> {
  const ip = req.ip ?? 'unknown'
  const now = Date.now()
  const failEntry = fails.get(ip)
  if (failEntry && now - failEntry.firstAt < FAIL_WINDOW_MS && failEntry.count >= FAIL_LIMIT) {
    res.header?.('Retry-After', String(Math.ceil((FAIL_WINDOW_MS - (now - failEntry.firstAt)) / 1000)))
    res.status(429).send('Too many login attempts. Try again later.')
    return
  }
  const password = String(req.body?.password ?? '')
  const next = getSafeNext(req.body?.next)
  if (!configuredPassword || !timingSafeEqual(password, configuredPassword)) {
    bumpFail(ip)
    res.set?.('Content-Type', 'text/html')
    res.status(401).send(renderFormShell('Relay login', renderLoginForm(next, 'Invalid password. Please try again.')))
    return
  }
  clearFail(ip)
  const sid = crypto.randomBytes(32).toString('hex')
  sessions.set(sid, { expiresAt: Date.now() + SESSION_TTL_MS })
  res.cookie('mcp_relay_session', sid, {
    httpOnly: true,
    secure: true,
    sameSite: 'lax',
    maxAge: SESSION_TTL_MS
  })
  res.redirect(next)
}
